Update smart contracts index.md (#7759)
ethereum/ethereum-org-website: src/content/smart-contracts/index.md

sidebar: true
---

> Smart contracts turn the terms of an agreement into computer code that executes when the contract terms are met.

Reviewer: My 2 gwei:

> turn the terms of an agreement into computer code

I don't like the framing here. It makes it sound like smart contract are primarily for legal contracts, which is a misnomer. They're computer programs first & foremost, which are capable of more broader functionality than merely turning terms of an "agreement" into computer code. There doesn't have to be any "agreement" between parties. e.g. is a smart contract wallet "terms of an agreement"? Maybe technically but I don't think it's a straightforward explanation. I feel this definition puts an inaccurate limit on what they're capable of.

> Smart contracts turn the terms of an agreement into computer code that executes when the contract terms are met.

Reviewer: > code that executes when the contract terms are met

This is another inaccuracy (which to be fair, existed before your PR). Smart contract code doesn't automatically execute "when the contract terms are met". Contracts still must be "poked" by an EOA in order to execute. e.g. if a loan in MakerDAO becomes undercollateralized (say due to a drop in the market price of ETH), the vault is not automatically liquidated. It requires someone to trigger a function on the smart contract - that's why it's necessary for an entire [incentive structure for "keepers" to fulfill this role](https://docs.makerdao.com/keepers/auction-keepers).

Nitpicking a bit here but I think this is an important aspect of smart contracts that we shouldn't oversell. They're not completely autonomous in the sense that they won't automatically execute logic & state changes UNLESS they are interacted with.

Author: This _does_ seem important and I hadn't known that. What would you suggest instead?

Reviewer 2: Made some adjustments here to address these concerns (which I do agree with), and to get this over the finish line.

> A smart contract is a computer program following "if X then Y" logic, which means they behave exactly as programmed and are resistant to participants' reneging on the terms of the contract. They are the fundamental building blocks of Ethereum's application layer.

Reviewer: ```suggestion
A smart contract is a computer program that often follows an "if this then that" logic, which means they behave exactly as programmed and are resistant to participants' reneging on the terms of the contract. They are the fundamental building blocks of Ethereum's application layer.
```

I think spelling out "if this then that" is more common sense & will make translations more straightforward.

# Introduction to smart contracts {#introduction-to-smart-contracts}

> A smart contract is a digital contract that automatically executes the agreed on terms when its conditions are met. Smart contracts are the fundamental building blocks of [Ethereum applications](/dapps/). Fundamentally, they are computer programs following "if X then Y" logic, which means they behave exactly as programmed and are resistant to participants' reneging on the terms of the contract. 

Reviewer 3: I think @setgree's edit reads nicely, and frames the benefits of smart contracts in a context users are already familiar with (i.e., people reneging on contractual agreements). Also, the average reader learning about smart contracts will find the "If X then Y" logic easier to follow (at least, I did at the time).

> A smart contract is a digital contract that automatically executes the agreed on terms when its conditions are met. Smart contracts are the fundamental building blocks of [Ethereum applications](/dapps/). Fundamentally, they are computer programs following "if X then Y" logic, which means they behave exactly as programmed and are resistant to participants' reneging on the terms of the contract. 

Author: BTW first sentence can be split into to two, as so:
> Smart contracts are digital contracts on the blockchain. When a smart contract's terms are met, the contract automatically executes the outcome that the parties agreed to.

> A smart contract is a digital contract that automatically executes the agreed on terms when its conditions are met. Smart contracts are the fundamental building blocks of [Ethereum applications](/dapps/). Fundamentally, they are computer programs following "if X then Y" logic, which means they behave exactly as programmed and are resistant to participants' reneging on the terms of the contract. 

Reviewer 4: I prefer @minimalsm 's original version because it explains what smart contracts are in terms anyone can understand: building blocks, on-chain computer programs. The additional wordage in the newer version is mostly padding - does it really help to say that a "smart contract is a digital contract", for example?

How about the following compromise:

```suggestion
A smart contract is the fundamental building block of [Ethereum applications](/dapps/). Fundamentally, they are computer programs following "if X then Y" logic, which means they behave exactly as programmed and are resistant to participants' reneging on the terms of the contract. 
```

> A smart contract is a digital contract that automatically executes the agreed on terms when its conditions are met. Smart contracts are the fundamental building blocks of [Ethereum applications](/dapps/). Fundamentally, they are computer programs following "if X then Y" logic, which means they behave exactly as programmed and are resistant to participants' reneging on the terms of the contract. 

Author: I am not the expert here, and I will defer to your judgment; but in general, I think the lede sentence should literally say what the thing is, rather than the function it serves. For instance, the Wikipedia entry on smart contracts begins:

> A smart contract is a computer program or a transaction protocol that is intended to automatically execute, control or document legally relevant events and actions according to the terms of a contract or an agreement.

This seems like it's on the right track. The lede sentence I chose might not be exactly right, but I think that the general goal of beginning with a definition rather than context/implications is the right overall direction.

Just my 2c 😃

> In 1994, Nick Szazbo wrote [an introduction to the concept](https://www.fon.hum.uva.nl/rob/Courses/InformationInSpeech/CDROM/Literature/LOTwinterschool2006/szabo.best.vwh.net/smart.contracts.html) and, in 1996, [an exploration of what smart contracts could do](https://www.fon.hum.uva.nl/rob/Courses/InformationInSpeech/CDROM/Literature/LOTwinterschool2006/szabo.best.vwh.net/smart_contracts_2.html).

Author: Nick Szazbo -> Nick Szabo
(my bad)

In 1994, Nick Szazbo wrote [an introduction to the concept](https://www.fon.hum.uva.nl/rob/Courses/InformationInSpeech/CDROM/Literature/LOTwinterschool2006/szabo.best.vwh.net/smart.contracts.html) and, in 1996, [an exploration of what smart contracts could do](https://www.fon.hum.uva.nl/rob/Courses/InformationInSpeech/CDROM/Literature/LOTwinterschool2006/szabo.best.vwh.net/smart_contracts_2.html).

Nick Szabo envisioned a digital marketplace built on these automatic, cryptographically secure processes. A place where transactions and business functions can happen trustlessly — without intermediaries. Smart contracts on Ethereum put this vision into practice.
Szabo envisioned a digital marketplace where automatic, cryptographically-secure processes enable transactions and business functions to happen without trusted intermediaries. Smart contracts on Ethereum put this vision into practice.

## What are contracts? {#what-are-contracts}

You're probably thinking: _"I'm not a lawyer! Why would I care about contracts?"_. For most people, contracts bring to mind needlessly long terms and conditions agreements or boring legal documents.

> Contracts are just agreements. That is, any form of agreement can be encapsulated within the conditions of a contract. Verbal agreements or pen-and-paper contracts are acceptable for many things, but they aren't without flaws.

Author: A) I think you're assuming a bit much about what the audience is 'probably thinking' -- I personally think that most readers will be familiar with contracts and don't think of them as 'only for lawyers.' I recently signed a lease on an apartment, for example, without a lawyer. 

But also, aren't these developer docs? Developer docs don't need editorializing about the importance of the subject, IMO, which is effectively an attempt to sell them on why they should care about the subject. They've already come to the page to learn.

As to the second bit, I thought it was repetitive with other sections/unnecessary to the core idea. But I'm not dead set on that removal or anything 😃

> Contracts are just agreements. That is, any form of agreement can be encapsulated within the conditions of a contract. Verbal agreements or pen-and-paper contracts are acceptable for many things, but they aren't without flaws.

Reviewer 5: > A) I think we're assuming more than we need to about what the audience is 'probably thinking' -- I personally think that most readers will be familiar with contracts and don't think of them as 'only for lawyers.' I recently signed a lease on an apartment, for example. Contracts feel like just everyday life to me? IDK

This is fair. IIRC this was added after some minor user testing that indicated that users with little understanding of what a smart contract was even found the name ("smart contract") to be a little intimidating/corporate.

> But also, aren't these developer docs? Developer docs don't need to persuade the audience about why they should care about the subject, IMO, because they've already come to the page to learn.

No, these are what we call "Learn pages". Learn pages should be low on assumptions of knowledge, accessible to users without a strong technical understanding, and—as much as possible—without a need for an advanced understanding of the English language.

> As to the second bit, I thought it was repetitive with other sections/unnecessary to the core idea. But it's fine to include it 😃

Again, I'm in favour but I'd like some other opinions on this because I'm aware that I have some bias here.

> Contracts are just agreements. That is, any form of agreement can be encapsulated within the conditions of a contract. Verbal agreements or pen-and-paper contracts are acceptable for many things, but they aren't without flaws.

Author: Thanks for the clarification, happy to defer on this and all matters

> Contracts are just agreements. That is, any form of agreement can be encapsulated within the conditions of a contract. Verbal agreements or pen-and-paper contracts are acceptable for many things, but they aren't without flaws.

Reviewer 4: Again, I think a compromise between these two options is best- delete the first paragraph, keep the second, IMO.

> Contracts are just agreements. That is, any form of agreement can be encapsulated within the conditions of a contract. Verbal agreements or pen-and-paper contracts are acceptable for many things, but they aren't without flaws.

Reviewer 4: going to remove these sentences as per the original PR because I think the point is now made well in later sections.

This silly example illustrates the problem with any non-smart agreement. Even if the conditions of the agreement get met (i.e. you are the winner of the race), you must still trust another person to fulfill the agreement (i.e. payout on the bet).

## Smart contracts {#smart-contracts}

Smart contracts digitize agreements by turning the terms of an agreement into computer code that automatically executes when the contract terms are met.

> 

Reviewer 4: agree - I moved the definition to the very first line of the page.

- You select a product
- The vending machine displays the price
- You pay the price
- The vending machine verifies that you paid the right amount
- The vending machine gives you your item.

The vending machine will only dispense your desired product after all requirements are met. If you don't select a product or insert enough money, the vending machine won't give out your product.

### Automatic execution {#automation}

> The main virtue of a smart contract is that it automatically executes the outcome when its conditions are met. There is no need to wait for a human to execute the result. This removes the need for a  trusted intermediary.

Author: 1. smart contracts don't remove the need for *trust* en toto, i.e. you still need to trust the Ethereum network's stability and security 😃 . How about "This removes the need for interpersonal trust?"
2. virtue has a secondary meaning of "a good or useful quality of a thing, e.g. "Mike was extolling the virtues of the car," but it would be totally fine to replace it with 'advantage' or 'benefit'

> The main virtue of a smart contract is that it automatically executes the outcome when its conditions are met. There is no need to wait for a human to execute the result. This removes the need for a  trusted intermediary.

Reviewer 5: 1. I don't really disagree but doesn't this concept go against the idea of "Don't trust, Verify" 🤔. I suppose the idea is that you can verify the code instead of trusting a black box.
2. Yeah, I get that but we need to remember that we're serving a lot of non-native English speakers, so we prefer to tend towards easier-to-understand over purple prose.

> The main virtue of a smart contract is that it automatically executes the outcome when its conditions are met. There is no need to wait for a human to execute the result. This removes the need for a  trusted intermediary.

Reviewer 2: Agree that none of this is completely trust free; the key point is of course not needing to trust another human, and instead you must trust the code instead, which is open instead of in a black-box. 

Somewhat indifferent towards "virtue" but would probably err on the side of using a more direct terms here, similar to what it was.

> The main virtue of a smart contract is that it automatically executes the outcome when its conditions are met. There is no need to wait for a human to execute the result. This removes the need for a  trusted intermediary.

Reviewer 4: ```suggestion
The main benefit of a smart contract is that it deterministically executes unambiguous code when certain conditions are met. There is no need to wait for a humans to interpret or negotiate the result. This removes the need for trusted intermediaries.
```

> The main virtue of a smart contract is that it automatically executes the outcome when its conditions are met. There is no need to wait for a human to execute the result. This removes the need for a  trusted intermediary.

Author: what does the phrase 'ambiguous code' mean? Ambiguous as in, a human reader might interpret it one of a few ways? Because you could also say that the whole smart contract idea hinges on the code _not_ being ambiguous, at least, not to the computer. Otherwise the contract couldn't deterministically execute.

> The main virtue of a smart contract is that it automatically executes the outcome when its conditions are met. There is no need to wait for a human to execute the result. This removes the need for a  trusted intermediary.

Reviewer 4: it says "unambiguous".

> The main benefit of a smart contract is that it deterministically executes unambiguous code when certain conditions are met. There is no need to wait for a humans to interpret or negotiate the result. This removes the need for trusted intermediaries.

Reviewer 2: ```suggestion
The main benefit of a smart contract is that it deterministically executes unambiguous code when certain conditions are met. There is no need to wait for a human to interpret or negotiate the result. This removes the need for trusted intermediaries.
```

> For example, you could write a smart contract that holds funds in escrow for a child, allowing them to withdraw funds after a specific date. If they try to withdraw before that, the smart contract won't execute. Or you could write a contract that automatically gives you a digital version of a car's title when you pay the dealer.

Reviewer 5: Although not wrong, it might be a little unclear what 'that' refers to here.

```suggestion
For example, you could write a smart contract that holds funds in escrow for a child, allowing them to withdraw funds after a specific date. If they try to withdraw before that date, the smart contract won't execute. Or you could write a contract that automatically gives you a digital version of a car's title when you pay the dealer.
```

### Predictable outcomes {#predictability}

> Traditional contracts are risky because they rely on people to implement them. For example, two judges might interpret a contract differently, which could lead to inconsistent decisions and unequal outcomes. Smart contracts remove this possibility. Instead, smart contracts execute precisely based on the conditions written within the contract's code. This precision means that given the same circumstances, the smart contract will produce the same result.

Reviewer 5: Wow, the original was pretty clunky.

```suggestion
Traditional contracts are risky because they rely on people to implement them. For example, two judges might interpret a contract differently, leading to inconsistent decisions and unequal outcomes. Smart contracts remove this possibility. Instead, smart contracts execute precisely based on the conditions written within the contract's code. This precision means that given the same circumstances, the smart contract will produce the same result.
```

Reviewer 5: I feel like we can do better than 'risky' here

Author: I agree -- maybe
> Traditional contracts rely on people to implement them, which introduces a potential point of failure.

Or something like that

Reviewer 5: That works. 

I'm kinda thinking using the word 'humans' over 'people' might be good here but I also think it might be a little bit too detached. Any thoughts?

Author: 'humans' is good! My company (https://www.globalincomecoin.com/) uses the word human a lot instead of 'users' -- I think because users might have bad connotations? -- so that seems pretty natural to me

Reviewer 2: Good with this 👍 Similarly, I think use of "human" here is appropriate

Reviewer 4: ```suggestion
Traditional contracts are ambiguous because they rely on humans to interpret and implement them. For example, two judges might interpret a contract differently, which could lead to inconsistent decisions and unequal outcomes. Smart contracts remove this possibility. Instead, smart contracts execute precisely based on the conditions written within the contract's code. This precision means that given the same circumstances, the smart contract will produce the same result.
```

### Public record {#public-record}

Smart contracts are useful for audits and tracking. Since Ethereum smart contracts are on a public blockchain, anyone can instantly track asset transfers and other related information. For example, you can check to see that someone sent money to your address.

### Privacy protection {#privacy-protection}

Smart contracts also protect your privacy. Since Ethereum is a pseudonymous network (your transactions are tied publicly to a unique cryptographic address, not your identity), you can protect your privacy from observers.

### Visible terms {#visible-terms}

Finally, like traditional contracts, you can check what's in a smart contract before you sign it (or otherwise interact with it). A smart contract's transparency guarantees that anyone can scrutinize it.

## Smart contract use cases {#use-cases}

> Smart contracts can do essentially anything that a computer programs do.

Author: Last sentence should be
> Smart contracts can do essentially anything that computer programs can do.

Reviewer 2: ```suggestion
Smart contracts can do essentially anything that computer programs can do.
```
